src/neural/loader.cc

```diff
@@ -46,25 +50,89 @@ FloatVectors LoadFloatsFromFile(const std::string& filename) {
     } else {
       bytes_read += sz;
       buffer.resize(bytes_read);
-      // Add newline in the end for the case it was not there.
-      buffer.push_back('\n');
       break;
     }
   }
   gzclose(file);
 
+  return buffer;
+}
+
+void DenormLayer(const lc0::Weights_Layer& layer, FloatVectors& vecs) {
+  FloatVector v;
+  auto& buffer = layer.params();
+  auto data = reinterpret_cast<const std::uint16_t*>(buffer.data());
+  int n = buffer.length() / 2;
+  v.resize(n);
+  for (int i = 0; i < n; i++) {
+    v[i] = data[i] / float(0xffff);
+    v[i] *= layer.max_val() - layer.min_val();
+    v[i] += layer.min_val();
+  }
+  vecs.emplace_back(v);
+}
+
+void DenormConvBlock(const lc0::Weights_ConvBlock& conv, FloatVectors& vecs) {
+  DenormLayer(conv.weights(), vecs);
+  DenormLayer(conv.biases(), vecs);
+  DenormLayer(conv.bn_means(), vecs);
+  DenormLayer(conv.bn_stddivs(), vecs);
+}
+
+FloatVectors LoadFloatsFromPbFile(const std::string& buffer) {
+  auto net = lc0::Net();
+  FloatVectors vecs;
+  net.ParseFromString(buffer);
+
+  std::string min_version(std::to_string(net.min_version().major()) + ".");
+  min_version += std::to_string(net.min_version().minor()) + ".";
+  min_version += std::to_string(net.min_version().patch());
+
+  if (net.min_version().major() > LC0_VERSION_MAJOR)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+  if (net.min_version().minor() > LC0_VERSION_MINOR)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+  if (net.min_version().patch() > LC0_VERSION_PATCH)
+    throw Exception("Weights require at least lc0 version: " + min_version);
+
+  if (net.format().weights_encoding() != lc0::Format::LINEAR16)
+    throw Exception("Invalid weight encoding");
+
+  auto& w = net.weights();
+
+  DenormConvBlock(w.input(), vecs);
+
+  for (int i = 0, n = w.residual_size(); i < n; i++) {
+    DenormConvBlock(w.residual(i).conv1(), vecs);
+    DenormConvBlock(w.residual(i).conv2(), vecs);
+  }
+
+  DenormConvBlock(w.policy(), vecs);
+  DenormLayer(w.ip_pol_w(), vecs);
+  DenormLayer(w.ip_pol_b(), vecs);
+  DenormConvBlock(w.value(), vecs);
+  DenormLayer(w.ip1_val_w(), vecs);
+  DenormLayer(w.ip1_val_b(), vecs);
+  DenormLayer(w.ip2_val_w(), vecs);
+  DenormLayer(w.ip2_val_b(), vecs);
+
+  return vecs;
+}
+
+FloatVectors LoadFloatsFromFile(std::string* buffer) {
```

Comment:
It's better to make buffer parameter a const reference.
It's a really weird side effect of a function that it adds '\n'.

Probably better to remove that and add after the loop:
```cpp
  if (!line.empty()) {
       result.emplace_back();
       result.back().swap(line);
  }
```

Reply:
I don't see an easy fix for this. LoadFloatsFromFile is altering the buffer in various ways.

Reply:
Which ways? I only see it adds `"\n"` to the end of the buffer and doesn't do anything else.
4 lines that I wrote in my previous comment is everything that is needed to remove that.
Just remove that `+= "\n"`, and add those 4 lines after `for` loop, and I think it will work.

Reply:
It also adds \0 back into the buffer in some cases I guess.

Reply:
Indeed! Fine to keep it then.